src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.kafka.connect.transforms.Transformation;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ConcatFields<R extends ConnectRecord<R>> implements Transformation<R> {
+    private ConcatFieldsConfig config;
+    private static final Logger log = LoggerFactory.getLogger(ConcatFields.class);
+
+    @Override
+    public ConfigDef config() {
+        return ConcatFieldsConfig.config();
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        this.config = new ConcatFieldsConfig(configs);
+    }
+
+    @Override
+    public R apply(final R record) {
+        if (record.value() == null) {
+            throw new DataException("Value can't be null: " + record);
+        }
+        final R newRecord;
+        final List<Object> fieldValues = new ArrayList<>();
+
+        if (record.value() instanceof Struct) {
+            final Struct struct = (Struct) record.value();
+            final SchemaBuilder newSchema = SchemaBuilder.struct();
+            if (record.valueSchema() != null) {
+                record.valueSchema().fields().forEach(field -> newSchema.field(field.name(), field.schema()));
+            } else {
+                struct.schema().fields().forEach(field -> newSchema.field(field.name(), field.schema()));
+            }
+            newSchema.field(config.fieldName(), Schema.OPTIONAL_STRING_SCHEMA);
+            final Struct newStruct = new Struct(newSchema);
+            struct.schema().fields().forEach(field -> {
+                newStruct.put(field.name(), struct.get(field));
+            });
+            config.fieldNames().forEach(field -> {
+                try {
+                    if (struct.get(field) == null) {
+                        fieldValues.add(config.fieldReplaceMissing());
+                    } else {
+                        fieldValues.add(struct.get(field));
+                    }
+                } catch (final Exception e) {
+                    log.debug(field + " is missing, concat will use " + config.fieldReplaceMissing());
+                    fieldValues.add(config.fieldReplaceMissing());
+                }
+            });
+            newStruct.put(config.fieldName(),
+                fieldValues.stream().map(Object::toString).collect(Collectors.joining(config.delimiter())));
+            newRecord = createNewRecord(record, newSchema.build(), newStruct);
+        } else if (record.value() instanceof Map) {
+            final Map newValue = (Map) record.value();
+            config.fieldNames().forEach(field -> {
+                if (newValue.get(field) == null) {
+                    fieldValues.add(config.fieldReplaceMissing());
+                } else {
+                    fieldValues.add(newValue.get(field));
+                }
+            });
+            newValue.put(config.fieldName(),
+                fieldValues.stream().map(Object::toString).collect(Collectors.joining(config.delimiter())));
```

Comment:
Better to use for-loop here instead of streams. `JDK14` contains optimization for such code, but not `JDK11`, which is our default JDK

Reply:
Gotcha, I'll switch to the for loop syntax!

- [ ] change from streams to for loop